### PR TITLE
[Docs] userful hooks

### PR DESCRIPTION
**Pull request by chhluo**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add a tutorial to introduce useful hooks provided by MMDetection and MMCV.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/tutorials/test_results_submission.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Tutorial 11: Test Results Submission
+# Tutorial 12: Test Results Submission
```
Comment (hhaAndroid) [inline docs/en/tutorials/test_results_submission.md:1]: Need to be added to `docs/en/tutorials/index.rst`

File: docs/en/tutorials/useful_hook.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## CheckpointHook
+
+## EvalHook and DistEvalHook
+
+## ExpMomentumEMAHook and LinearMomentumEMAHook
+
+## NumClassCheckHook
+
+## [MemoryProfilerHook](https://github.com/open-mmlab/mmdetection/blob/dev/mmdet/core/hook/memory_profiler_hook.py)
```
Comment (jbwang1997) [inline docs/en/tutorials/useful_hook.md:15]: The link is pointing to dev branch, Need to change the master branch link.

File: docs/en/tutorials/useful_hook.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+## SyncNormHook
+
+## SyncRandomSizeHook
+
+## TextLoggerHook
+
+## TensorboardLoggerHook
+
+## WandbLogger
```
Comment (jbwang1997) [inline docs/en/tutorials/useful_hook.md:47]: MMDetWandbHook

**Review by jbwang1997 — APPROVED** [review]
LGTM

File: docs/en/tutorials/useful_hook.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Tutorial 13: Useful Hooks
+
+MMDetection and MMCV provide users with various useful hooks including log hooks, evaluation hooks, NumClassCheckHook and etc. . This tutorial mainly introduce function and usage of hooks in MMDetection\. For how to add custom hook to MMDetection, please read [customize_runtime](https://mmdetection.readthedocs.io/en/latest/tutorials/customize_runtime.html#customize-self-implemented-hooks),  and for hooks in MMCV, please read [docs in MMCV](https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/runner.md).
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hook.md:3]: - 'NumClassCheckHook and etc. .' -> 'NumClassCheckHook, etc.'
- mainly introduce -> introduces the functionalities and usages of hooks implemented in MMDetection.
- We should add a brief example here to indicate how to add a custom hook, then refer readers to the doc for more details.
- and for hooks in MMCV -> For using hooks in MMCV, please read the API documentation in MMCV.

File: docs/en/tutorials/useful_hook.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+## EvalHook and DistEvalHook
+
+## ExpMomentumEMAHook and LinearMomentumEMAHook
+
+## NumClassCheckHook
+
+## [MemoryProfilerHook](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/hook/memory_profiler_hook.py)
+
+Memory profiler hook records memory information including virtual memory, swap memory, and the memory of the current process. This hook can help us grasp the memory usage of the system and discover potential memory leak bugs. To use this hook, we must install `memory_profiler` and `psutil` by `pip install memory_profiler psutil` first.
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hook.md:15]: - This hook can help us grasp the -> This hook helps grasp the
- To use this hook, users should install

File: docs/en/tutorials/useful_hook.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+```python
+custom_hooks = [
+    dict(type='MemoryProfilerHook', interval=50)
+]
+```
+
+### Result
+
+The following is the result of `MemoryProfilerHook`.
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hook.md:29]: Should be more specific:
During training, you can see the messages in the log recorded by `Mexxx` as below

**Review by ZwwWayne — APPROVED** [review]

File: docs/en/tutorials/useful_hooks.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+## SyncRandomSizeHook
+
+## YOLOXLrUpdaterHook
+
+## YOLOXModeSwitchHook
+
+
+## How to implement a custom hook
+
+In general, there are 10 sites where hooks can be inserted from the beginning to the end of model training. The users can implement custom hooks and inserted them in different sites in the process of training to do what they want.
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hooks.md:49]:
```suggestion
In general, there are 10 places where hooks can be inserted from the beginning to the end of model training. The users can implement custom hooks and insert them at different places in the process of training to do what they want.
```

File: docs/en/tutorials/useful_hooks.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+- global sites: `before_run`, `after_run`
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hooks.md:51]: sites -> points

File: docs/en/tutorials/useful_hooks.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+- sites in training: `before_train_epoch`, `before_train_iter`, `after_train_iter`, `after_train_epoch`
+- sites in validation: `before_val_epoch`, `before_val_iter`, `after_val_iter`, `after_val_epoch`
+
+Loss going NaN is a common penomenon, users can implement a hook to help check loss and terminate training when loss goes NaN. To implement such a hook, the users should base on Hook provided in MMCV, and implement `after_train_iter` method which checks that whether loss goes NaN after every n training iterations. After that,  register it in `HOOKS`. At Last, add `custom_hooks = [dict(type='MemoryProfilerHook', interval=50)]` in the config file. Please read [customize_runtime](https://mmdetection.readthedocs.io/en/latest/tutorials/customize_runtime.html#customize-self-implemented-hooks) for more about implementing a custom hook.
```
Comment (ZwwWayne) [inline docs/en/tutorials/useful_hooks.md:55]:
```suggestion
For example, users can implement a hook to check loss and terminate training when loss goes NaN. To achieve that, there is three steps to go:

1. Implement a new hook that inherits the `Hook` class in MMCV, and implement `after_train_iter` method which checks whether loss goes NaN after every `n` training iteration. 
2. The implemented hook should be registered in `HOOKS` by `@HOOKS.register_module()` as shown in the code below.
3. Add `custom_hooks = [dict(type='MemoryProfilerHook', interval=50)]` in the config file.

Please read [customize_runtime](https://mmdetection.readthedocs.io/en/latest/tutorials/customize_runtime.html#customize-self-implemented-hooks) for more about implementing a custom hook.
```

**Pull request merged**